<a href="https://colab.research.google.com/github/amritaa72/Runs-and-Wickets/blob/main/ball_by_ball_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount("/content/drive")

Mounted at /content/drive


In [3]:
data=pd.read_csv("/content/drive/MyDrive/Cricket_pandas/virat_kohli_ball_by_ball_2022.csv")
data

,ID,innings,overs,ballnumber,batter,bowler,non-striker,extra_type,batsman_run,extras_run,total_run,non_boundary,isWicketDelivery,player_out,kind,fielders_involved,BattingTeam
0,1304049,1,7,2,V Kohli,Harpreet Brar,F du Plessis,NaN,1,0,1,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
1,1304049,1,7,4,V Kohli,Harpreet Brar,F du Plessis,NaN,0,0,0,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
2,1304049,1,7,5,V Kohli,Harpreet Brar,F du Plessis,NaN,1,0,1,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
3,1304049,1,8,1,V Kohli,RD Chahar,F du Plessis,NaN,1,0,1,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
4,1304049,1,9,1,V Kohli,Harpreet Brar,F du Plessis,NaN,2,0,2,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
301,1312199,1,0,5,V Kohli,TA Boult,F du Plessis,NaN,0,0,0,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
302,1312199,1,0,6,V Kohli,TA Boult,F du Plessis,NaN,6,0,6,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
303,1312199,1,1,3,V Kohli,M Prasidh Krishna,F du Plessis,NaN,0,0,0,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
304,1312199,1,1,4,V Kohli,M Prasidh Krishna,F du Plessis,NaN,0,0,0,0,0,NaN,NaN,NaN,Royal Challengers Bangalore


In [4]:
data.columns

Index(['ID', 'innings', 'overs', 'ballnumber', 'batter', 'bowler',
       'non-striker', 'extra_type', 'batsman_run', 'extras_run', 'total_run',
       'non_boundary', 'isWicketDelivery', 'player_out', 'kind',
       'fielders_involved', 'BattingTeam'],
      dtype='object')

In [5]:
data.isna()

,ID,innings,overs,ballnumber,batter,bowler,non-striker,extra_type,batsman_run,extras_run,total_run,non_boundary,isWicketDelivery,player_out,kind,fielders_involved,BattingTeam
0,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
1,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
2,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
3,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
4,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
301,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
302,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
303,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
304,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False


In [6]:
data.isna().sum()

,0
ID,0
innings,0
overs,0
ballnumber,0
batter,0
bowler,0
non-striker,0
extra_type,292
batsman_run,0
extras_run,0


In [7]:
data.head()

,ID,innings,overs,ballnumber,batter,bowler,non-striker,extra_type,batsman_run,extras_run,total_run,non_boundary,isWicketDelivery,player_out,kind,fielders_involved,BattingTeam
0,1304049,1,7,2,V Kohli,Harpreet Brar,F du Plessis,NaN,1,0,1,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
1,1304049,1,7,4,V Kohli,Harpreet Brar,F du Plessis,NaN,0,0,0,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
2,1304049,1,7,5,V Kohli,Harpreet Brar,F du Plessis,NaN,1,0,1,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
3,1304049,1,8,1,V Kohli,RD Chahar,F du Plessis,NaN,1,0,1,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
4,1304049,1,9,1,V Kohli,Harpreet Brar,F du Plessis,NaN,2,0,2,0,0,NaN,NaN,NaN,Royal Challengers Bangalore


In [8]:
data.tail()

,ID,innings,overs,ballnumber,batter,bowler,non-striker,extra_type,batsman_run,extras_run,total_run,non_boundary,isWicketDelivery,player_out,kind,fielders_involved,BattingTeam
301,1312199,1,0,5,V Kohli,TA Boult,F du Plessis,NaN,0,0,0,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
302,1312199,1,0,6,V Kohli,TA Boult,F du Plessis,NaN,6,0,6,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
303,1312199,1,1,3,V Kohli,M Prasidh Krishna,F du Plessis,NaN,0,0,0,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
304,1312199,1,1,4,V Kohli,M Prasidh Krishna,F du Plessis,NaN,0,0,0,0,0,NaN,NaN,NaN,Royal Challengers Bangalore
305,1312199,1,1,5,V Kohli,M Prasidh Krishna,F du Plessis,NaN,0,0,0,0,1,V Kohli,caught,SV Samson,Royal Challengers Bangalore


In [9]:
data.shape

(306, 17)

**306 rows** and **17 columns**

essentially Virat has **played 306 balls** in that seasson (2022)

In [10]:
data.columns

Index(['ID', 'innings', 'overs', 'ballnumber', 'batter', 'bowler',
       'non-striker', 'extra_type', 'batsman_run', 'extras_run', 'total_run',
       'non_boundary', 'isWicketDelivery', 'player_out', 'kind',
       'fielders_involved', 'BattingTeam'],
      dtype='object')

In [11]:
data.rows

AttributeError: 'DataFrame' object has no attribute 'rows'

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306 entries, 0 to 305
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   ID                 306 non-null    int64 
 1   innings            306 non-null    int64 
 2   overs              306 non-null    int64 
 3   ballnumber         306 non-null    int64 
 4   batter             306 non-null    object
 5   bowler             306 non-null    object
 6   non-striker        306 non-null    object
 7   extra_type         14 non-null     object
 8   batsman_run        306 non-null    int64 
 9   extras_run         306 non-null    int64 
 10  total_run          306 non-null    int64 
 11  non_boundary       306 non-null    int64 
 12  isWicketDelivery   306 non-null    int64 
 13  player_out         16 non-null     object
 14  kind               16 non-null     object
 15  fielders_involved  13 non-null     object
 16  BattingTeam        306 non-null    object
dt

In [13]:
data.describe()

,ID,innings,overs,ballnumber,batsman_run,extras_run,total_run,non_boundary,isWicketDelivery
count,3.060000e+02,306.000000,306.00000,306.000000,306.000000,306.000000,306.000000,306.0,306.000000
mean,1.304986e+06,1.441176,6.46732,3.382353,1.114379,0.068627,1.183007,0.0,0.052288
std,2.554006e+03,0.497341,5.12025,1.744560,1.454216,0.411151,1.459566,0.0,0.222971
min,1.304049e+06,1.000000,0.00000,1.000000,0.000000,0.000000,0.000000,0.0,0.000000
25%,1.304064e+06,1.000000,2.00000,2.000000,0.000000,0.000000,0.000000,0.0,0.000000
50%,1.304089e+06,1.000000,6.00000,3.000000,1.000000,0.000000,1.000000,0.0,0.000000
75%,1.304113e+06,2.000000,10.00000,5.000000,1.000000,0.000000,1.000000,0.0,0.000000
max,1.312199e+06,2.000000,19.00000,7.000000,6.000000,5.000000,6.000000,0.0,1.000000


**DATA CLEANING**

In [14]:
data.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
301,False
302,False
303,False
304,False


In [15]:
data[data.duplicated()].sum()

,0
ID,0
innings,0
overs,0
ballnumber,0
batter,0
bowler,0
non-striker,0
extra_type,0
batsman_run,0
extras_run,0


In [16]:
data.isna()

,ID,innings,overs,ballnumber,batter,bowler,non-striker,extra_type,batsman_run,extras_run,total_run,non_boundary,isWicketDelivery,player_out,kind,fielders_involved,BattingTeam
0,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
1,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
2,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
3,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
4,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
301,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
302,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
303,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False
304,False,False,False,False,False,False,False,True,False,False,False,False,False,True,True,True,False


In [17]:
data.isna().sum()

,0
ID,0
innings,0
overs,0
ballnumber,0
batter,0
bowler,0
non-striker,0
extra_type,292
batsman_run,0
extras_run,0




*   This dataset is free from reddundant records or null values

*   In case of duplicate values we drop them

*   In case of null values, we fill it using fillna() depending on the data

*   extra_type, player_out, kind, fielders_invloved have null values, because we have these values only when the bowler bowls extras (wide,no_ball), take wicket.

*   We are aware that these are deliberately missing, and arent missing due to other reasons









### **Data Analysis**

Total runs scored by Virat Kohli in 2022

In [18]:
#calculate total runs scored by virat kohli in 2022
total_runs=data['batsman_run'].sum()
print("Total Runs scored:", total_runs)


Total Runs scored: 341


Number of fours scored by VK

In [19]:
#calculate the total number of 4s scored by virat Kohli
fours= (data['batsman_run']==4).sum()
print("Total 4s:  ",fours)

Total 4s:   32


Number of sixes hit by VK

In [20]:
#calculate the toal 6s scored by VK
sixes=(data['batsman_run']==4).sum()
print("Total 6s   :" ,sixes)

Total 6s   : 32


TOTAL BOUNDARIES

Total boundaries= Total 6s+ Total 4s

In [21]:
#calculate total boundaroes hit
boundary= ((data["batsman_run"]==4) or (data['batsman_run']==6)).sum()
print("total boundaries :",boundary)

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [22]:
#total boundaries hit
total_boundaries= fours+sixes
print("Total boundaries :  ",total_boundaries)


Total boundaries :   64


STRIKE RATE OF VK

In [23]:
extras= data["extra_type"].value_counts()
extras

,count
extra_type,
wides,12
legbyes,2


In [24]:
total_balls_faced=len(data['extra_type'])-(data['extra_type']=='wides').sum()
print("Total balls faced:  ", total_balls_faced)

Total balls faced:   294


# **Strike Rate**

strike rate=( **[**total runs scored by batsman]**/** **[  total balls faced by batsman]** ) x **100**



In [25]:
strike_rate= (total_runs/total_balls_faced)*100
print("Strike Rate : ", strike_rate)

Strike Rate :  115.98639455782313


Number of **SINGLES**

In [26]:
#number of singles scored by VK
singles=(data['batsman_run']==1).sum()
print("Total 1s : ",singles)

Total 1s :  117


In [27]:
#number of doubles by VK
doubles= (data['batsman_run']==2).sum()
print(" Total 2s :", doubles)

 Total 2s : 18


batting avg = **total runs scored/ times out**

In [28]:
#times out
times_out=(data[data['player_out']=='V Kohli']['isWicketDelivery']).sum()
#or just (data[data['player_out']=='V Kohli']).sum()
print("Number of times got out :", times_out)

Number of times got out : 14


In [29]:
#batting avg
batting_avg= (total_runs/times_out)
print(f"Batting Avg : { batting_avg: .2f}")

Batting Avg :  24.36


# **Most common over Virat Got out**

In [30]:
most_common_over=data[data['player_out']=='V Kohli']['overs'].value_counts()
most_common_over

,count
overs,
1,3
0,2
16,2
2,1
18,1
6,1
4,1
9,1
3,1


In [33]:
most_common_over=data[data['player_out']=='V Kohli']['overs'].value_counts().idxmax()
print("Most common over : " ,most_common_over)


Most common over :  1


In [38]:
#Which over does he get out most of the time?
common_over= data[data['player_out']=='V Kohli']['overs'].value_counts()
common_over
percentage_common_over=[]
for i in common_over:
  percentage_over=(i/times_out)*100
  percentage_common_over.append(percentage_over)
print(f"Percentage common over : {percentage_common_over[0]}, in over {common_over.idxmax()}")


Percentage common over : 21.428571428571427, in over 1


Every 5th match kinda getting out in the first over of the game

# **Which kind of dissmissal did kohli get out**

In [46]:
#which kind of dissmissal did Kohli get out most of the time
common_dissmissal= data[data['player_out']=='V Kohli']['kind'].value_counts()
common_dissmissal
print("Most common dissmissal :", common_dissmissal)

Most common dissmissal : kind
caught     9
bowled     2
lbw        1
run out    1
stumped    1
Name: count, dtype: int64


In [47]:
#percentage of the kind through which he was ot most times
percentage_common_dissmissal=[]
for j in common_dissmissal:
  percentage_over=(j/times_out)*100
  percentage_common_dissmissal.append(percentage_over)
print(f"Percentage common dissmissal : {percentage_common_dissmissal[0]}, of kind {common_dissmissal.idxmax()}")


Percentage common dissmissal : 64.28571428571429, of kind caught


# **Batting performance during powerplay**

In [49]:
#phase of innings
powerplay_overs=range(0,7)
middle_overs=range(7,16)
death_overs=range(17,21)

#feature engg
#create column innings_phase to idesntify phase of each delivery
def categorise_innings_phase(overs):
  if overs in powerplay_overs:
    return 'Powerplay'
  elif overs in middle_overs:
    return 'Middle Overs'
  elif overs in death_overs:
    return 'Death overs'
  else:
    return 'other'

data['innings_phase']=data['overs'].apply(categorise_innings_phase)
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306 entries, 0 to 305
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   ID                 306 non-null    int64 
 1   innings            306 non-null    int64 
 2   overs              306 non-null    int64 
 3   ballnumber         306 non-null    int64 
 4   batter             306 non-null    object
 5   bowler             306 non-null    object
 6   non-striker        306 non-null    object
 7   extra_type         14 non-null     object
 8   batsman_run        306 non-null    int64 
 9   extras_run         306 non-null    int64 
 10  total_run          306 non-null    int64 
 11  non_boundary       306 non-null    int64 
 12  isWicketDelivery   306 non-null    int64 
 13  player_out         16 non-null     object
 14  kind               16 non-null     object
 15  fielders_involved  13 non-null     object
 16  BattingTeam        306 non-null    object
 1

In [53]:
phase_group=data.groupby('innings_phase')

# calculate various batting performance metric for each phase
phase_stats=phase_group.agg({'total_run':'sum','batsman_run':'sum','ballnumber':'count'})
phase_stats

,total_run,batsman_run,ballnumber
innings_phase,,,
Death overs,9,9,8
Middle Overs,159,151,129
Powerplay,181,168,158
other,13,13,11


In [64]:
powerplay_overs= range(0,7)
middle_overs=range(7,16)
death_overs=range(16,21)

#filter for powerplay, middle overs and death overs
powerplay_df=data[data['overs'].isin(powerplay_overs)]
middleovers_df=data[data['overs'].isin(middle_overs)]
deathovers_df=data[data['overs'].isin(death_overs)]

#calculate various batting performance metrics for each phase
powerplay_stats=powerplay_df.agg({'total_run':'sum','batsman_run':'sum',"ballnumber":'count'})
middle_over_stats=middleovers_df.agg({'total_run':'sum','batsman_run':'sum',"ballnumber":'count'})
deathover_stats=deathovers_df.agg({'total_run':'sum','batsman_run':'sum',"ballnumber":'count'})

powerplay_stats['strike_rate']=(powerplay_stats['batsman_run']/powerplay_stats['ballnumber'])*100
middle_over_stats['strike_rate']=(middle_over_stats['batsman_run']/middle_over_stats['ballnumber'])*100
deathover_stats['strike_rate']=(deathover_stats['batsman_run']/deathover_stats['ballnumber'])*100

powerplay_stats
#middle_over_stats
#deathover_stats



,0
total_run,159.000000
batsman_run,151.000000
ballnumber,129.000000
strike_rate,117.054264


# **statistics to dataframe format and transpose**

In [65]:
power_play_stats_df=pd.DataFrame(powerplay_stats)
power_play_stats_df

,0
total_run,181.000000
batsman_run,168.000000
ballnumber,158.000000
strike_rate,106.329114


In [76]:
power_play_stats_df=pd.DataFrame(powerplay_stats).transpose()
power_play_stats_df.columns=['total_run'," kohli_runs",'balls_faced',"strike_rate"]

middle_overs_stats_df=pd.DataFrame(middle_over_stats).transpose()
middle_overs_stats_df.columns=['total_run'," kohli_runs",'balls_faced',"strike_rate"]

death_overs_stats_df=pd.DataFrame(deathover_stats).transpose()
death_overs_stats_df.columns=['total_run'," kohli_runs",'balls_faced',"strike_rate"]

death_overs_stats_df


,total_run,kohli_runs,balls_faced,strike_rate
0,22.0,22.0,19.0,115.789474


In [77]:

#display results!!
print('POWER PLAY STATISTICS!')
print(power_play_stats_df)
print("\n MIDDLE OVERS STATISTICS")
print(middle_overs_stats_df)
print("\n DEATH OVERS STATISTICS")
print(death_overs_stats_df)

POWER PLAY STATISTICS!
   total_run   kohli_runs  balls_faced  strike_rate
0      181.0        168.0        158.0   106.329114

 MIDDLE OVERS STATISTICS
   total_run   kohli_runs  balls_faced  strike_rate
0      159.0        151.0        129.0   117.054264

 DEATH OVERS STATISTICS
   total_run   kohli_runs  balls_faced  strike_rate
0       22.0         22.0         19.0   115.789474


### Converting float columns to integers


In [78]:
power_play_stats_df[['total_run'," kohli_runs",'balls_faced',"strike_rate"]]=power_play_stats_df[['total_run'," kohli_runs",'balls_faced',"strike_rate"]].astype(int)
middle_overs_stats_df[['total_run'," kohli_runs",'balls_faced',"strike_rate"]]=middle_overs_stats_df[['total_run'," kohli_runs",'balls_faced',"strike_rate"]].astype(int)
death_overs_stats_df[['total_run'," kohli_runs",'balls_faced',"strike_rate"]]=death_overs_stats_df[['total_run'," kohli_runs",'balls_faced',"strike_rate"]].astype(int)

#display results!!
print('POWER PLAY STATISTICS!')
print(power_play_stats_df)
print("\n MIDDLE OVERS STATISTICS")
print(middle_overs_stats_df)
print("\n DEATH OVERS STATISTICS")
print(death_overs_stats_df)

POWER PLAY STATISTICS!
   total_run   kohli_runs  balls_faced  strike_rate
0        181          168          158          106

 MIDDLE OVERS STATISTICS
   total_run   kohli_runs  balls_faced  strike_rate
0        159          151          129          117

 DEATH OVERS STATISTICS
   total_run   kohli_runs  balls_faced  strike_rate
0         22           22           19          115


# CONCLUSION
After analysing the ball by all data of Virat Kohli during IPL season 2022
The conclusions are:


*   **Total average** : Virat Kohli scored **341(294)** runs in IPL 2022
*   **Strike rate** : Virat Maintained a strike rate of **11.7.054**
*   **Batting average**: **24.36** in IPL 2022
*   He had a **slower strike rate** during the **powerplay overs**, was **inconsistent** through runs in IPL 2022
*   Most **commonly dissmissed** through getting **caught** : **64.29%**
*   Faced **Higher risk of getting out** in the **first over** : **21.43%** and especially on the first ball
*   Manage to maintain **decent overall strike rate** in **middle overs** and **contributed significantly** with **40 boundaries** to his total 341 runs in IPL 2022
*   One **major issue** for Kohli is **low strike rate during power-play**.
*   Instead of finding the right balance between boundaries and singles, he **settled more for singles(117)**
*   Overall, it was a **challenging season for Kohli in IPL 2022**.









